// just-task/lib/enableTypeScript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resolve_1 = require("./resolve");
const just_task_logger_1 = require("just-task-logger");
function enableTypeScript({ transpileOnly = true }) {
    const tsNodeModule = resolve_1.resolve('ts-node');
    if (tsNodeModule) {
        const tsNode = require(tsNodeModule);
        tsNode.register({
            transpileOnly,
            skipProject: true,
            compilerOptions: {
                target: 'esnext',
                module: 'commonjs',
                strict: false,
                skipLibCheck: true,
                skipDefaultLibCheck: true,
                moduleResolution: 'node',
                allowJs: true,
                esModuleInterop: true
            },
            files: ['just.config.ts']
        });
    }
    else {
        just_task_logger_1.logger.error(`In order to use TypeScript with just.config.ts, you need to install "ts-node" module:

  npm install -D ts-node

or

  yarn add -D ts-node

`);
    }
}
exports.enableTypeScript = enableTypeScript;
