// just-task-logger/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const yargs_1 = require("yargs");
const perf_1 = require("./perf");
function logInternal(method, symbol, ...args) {
    const now = new Date();
    const timestamp = chalk_1.default.gray(`[${now.toLocaleTimeString()}]`);
    console[method](timestamp, symbol, ...args);
}
const emptySquare = '\u25a1';
const square = '\u25a0';
const triangle = '\u25b2';
exports.logger = {
    enableVerbose: !!yargs_1.argv.verbose,
    verbose(...args) {
        if (exports.logger.enableVerbose) {
            logInternal('info', chalk_1.default.gray(emptySquare), ...args);
        }
    },
    info(...args) {
        logInternal('info', chalk_1.default.green(square), ...args);
    },
    warn(...args) {
        logInternal('warn', chalk_1.default.yellow(triangle), ...args);
    },
    error(...args) {
        logInternal('error', chalk_1.default.redBright('x'), ...args);
    },
    perf(marker, ...args) {
        if (exports.logger.enableVerbose) {
            const delta = perf_1.getDeltaAndClearMark(marker);
            if (delta) {
                const ns = delta[0] * 1e9 + delta[1];
                const deltaMsg = `${ns / 1e9}s`;
                logInternal('info', chalk_1.default.cyan(square), `mark(${chalk_1.default.cyanBright(marker)}): took ${chalk_1.default.cyanBright(deltaMsg)}`, ...args);
            }
        }
    }
};
